secret-memory/src/allocator.rs

```diff
@@ -0,0 +1,239 @@
+            AllocInner::SodiumAlloc(_) => fmt.write_str("<memsec based Rust allocator>")
+        }
+    }
+}
+
+#[cfg(not(unix))]
+fn get_support_memfd_secret() -> bool {
+    false
+}
+
+#[cfg(unix)]
+fn get_support_memfd_secret() -> bool {
+    static SUPPORT_MEMFD_SECRET: std::sync::OnceLock<bool> = std::sync::OnceLock::new();
+    *SUPPORT_MEMFD_SECRET.get_or_init(|| {
+        let fd  = unsafe {
+            libc::syscall(libc::SYS_memfd_secret, 0) as i32
+        };
+        if fd != -1 || if let Some(os_err) = Error::last_os_error().raw_os_error() { os_err != 38 } else { true } {
+            // with memfd_secret support
+            return true;
+        }
+        if fd != -1 {
+            unsafe {
+                libc::close(fd);
+            }
+        }
+        false
+    })
+}
+
+#[derive(Clone, Default)]
+struct SodiumAlloc;
+
+impl SodiumAlloc {
+    fn do_allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        // Call sodium allocator
+        let ptr = unsafe { memsec::malloc_sized(layout.size()) };
+
+        let ptr = if let Some(p) = ptr {
+            p
+        } else {
+            log::error!(
+                "Allocation {layout:?} was requested but memsec returned a null pointer"
+            );
+            return Err(AllocError);
+        };
+
+        // Ensure the right allocation is used
+        let off = (ptr.as_ptr() as *const libc::c_void).align_offset(layout.align());
+        if off != 0 {
+            log::error!("Allocation {layout:?} was requested but memsec returned allocation \
+                with offset {off} from the requested alignment. Memsec always allocates values \
+                at the end of a memory page for security reasons, custom alignments are not supported. \
+                You could try allocating an oversized value.");
+            return Err(AllocError);
+        }
+
+        Ok(ptr)
+    }
+
+    fn do_deallocate(&self, ptr: NonNull<u8>, _layout: Layout) {
+        unsafe {
+            memsec::free(ptr)
+        }
+    }
+}
+
+
+#[derive(Clone)]
+struct SecretAlloc {
+    fd : i32,
+}
+
+impl SecretAlloc {
+    pub fn new() -> Self {
+        let fd  = unsafe {
+            libc::syscall(libc::SYS_memfd_secret, 0) as i32
+        };
+        if fd == -1 {
+            log::error!(
+                "Create secret file descriptor failed {}.",
+                std::io::Error::last_os_error()
+            );
+        }
+        SecretAlloc {
+            fd: fd,
+        }
+    }
+
+    fn do_allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        if self.fd == -1 {
+            log::error!(
+                "Create secret file descriptor failed."
+            );
+            return Err(AllocError);
+        }
+        let ptr = unsafe {
+            let ret = libc::mmap(null_mut::<libc::c_void>(), layout.size(), libc::PROT_READ | libc::PROT_WRITE, libc::MAP_LOCKED, self.fd, 0);
```

Comment:
This is missing the MAP_PRIVATE or MAP_SHARED flag, so this will always fail:

> The flags argument determines whether updates to the mapping are
>       visible to other processes mapping the same region, and whether updates
>       are carried through to the underlying file.  This behavior is
>       determined by including exactly one of the following values in flags: